[[Go]]
Мы можем создавать каастомные ошибки в виде структур. Главное, чтобы эта структура удовлетворяла интерфейсу `error`
```go
type MyCustomError struct {
    text string
    code int
}

func (e MyCustomError) Error() string {
    return e.text
}

func handleHttp(path string) error {
    if path != "/valid_path" {
        return MyCustomError{
            text: "not found",
            code: 404,
        }
    }
    return nil
}

func main() {
    err := handleHttp("qwe")
    if err != nil {
        customErr, ok := err.(MyCustomError) // type assert
        if ok {
            fmt.Println(customErr.code, customErr.text)
        } else {
            fmt.Println(err.Error())
        }
    }
}
```
Мы проверяем, что ошибка, которая прилетела является нашем типом ошибки. Но так писать не надо, так как для этого есть встроенные функции
### `errors.As`
Функция `errors.As` делает как раз то, что написано выше - приводит ошибку к какой-то конкретной структуре
```go
func main() {
    err := handleHttp("qwe")
    if err != nil {
        var myErr MyCustomError
        if errors.As(err, &myErr) {
            fmt.Println(myErr.code, myErr.text)
        } else {
            fmt.Println(err.Error())
        }
    }
}
```
Теперь понятно, почему туда надо передавать указатель

## `errors.Is`
Проверяет, нет ли в цепочке ошибок нужной нам
```go
var Err = errors.New("GLOBAL ERROR")
func main() {
    err := superup()
    if errors.Is(err, Err) {
        fmt.Println("ok")       // ok
        fmt.Println(err == Err) // false
    }
}

func superup() error {
    return fmt.Errorf("in superup: %w", Err)
}
```
Мы не можем сравнивать через `==`, так как могут быть обернутые ошибки. При оборачивании старая ошибка не удаляется, а просто сохранятся внутри. 